### PR TITLE
Add iou loss

### PR DESCRIPTION
**Pull request by shanibenbaruch**
1. create structure loss - abstract class 
2. split dice loss + iou loss to functions that inherit from structure loss
3. add iou loss test

### CODE REVIEW

**Review by lkdci — CHANGES_REQUESTED** [review]
Great job, few comments

File: tests/unit_tests/iou_loss_test.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
+import torch
+import unittest
+
+from super_gradients.training.losses.iou_loss import IoULoss, GeneralizedIoULoss, BinaryIoULoss
+
+
+class IoULossTest(unittest.TestCase):
```
[the remark this answers is not in the document]
Comment (shanibenbaruch) [reply tests/unit_tests/iou_loss_test.py:7]: Added in 1d42d9558a3eea9b2cc5d1aa1b9d027522e3000b

File: src/super_gradients/training/losses/iou_loss.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    Compute IoU Loss for binary class tasks (1 class only).
+    Except target to be a binary map with 0 and 1 values.
+    """
+    def __init__(self,
+                 apply_sigmoid: bool = True,
+                 smooth: float = 1.,
+                 eps: float = 1e-5):
+        """
+        :param apply_sigmoid: Whether to apply sigmoid to the predictions.
+        :param smooth: laplace smoothing, also known as additive smoothing. The larger smooth value is, closer the IoU
```
[the remark this answers is not in the document]
Comment (shanibenbaruch) [reply src/super_gradients/training/losses/iou_loss.py:53]: removed in 1d42d9558a3eea9b2cc5d1aa1b9d027522e3000b

File: src/super_gradients/training/losses/dice_loss.py [hunk 3]
```diff
@@ -1,99 +1,43 @@
...
-
-        reduce_spatial_dims = list(range(2, len(predict.shape)))
-        reduce_dims = [1] + reduce_spatial_dims if self.sum_over_batches else [0] + reduce_spatial_dims
-
-        intersection = torch.sum(labels_one_hot * predict, dim=reduce_dims)
+        :param labels_one_hot: target in one hot format.   shape: [BS, num_classes, img_width, img_height]
+        :param predict: predictions tensor.                shape: [BS, num_classes, img_width, img_height]
+        :return:
+            numerator = intersection between predictions and target. shape: [BS, num_classes, img_width, img_height]
+            denominator = sum of predictions and target.             shape: [BS, num_classes, img_width, img_height]
```
[the remark this answers is not in the document]
Comment (shanibenbaruch) [reply src/super_gradients/training/losses/dice_loss.py]: added in 1d42d9558a3eea9b2cc5d1aa1b9d027522e3000b

File: src/super_gradients/training/losses/structure_loss.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+from super_gradients.common.abstractions.abstract_logger import get_logger
+from super_gradients.training.losses.loss_utils import apply_reduce, LossReduction
+from super_gradients.training.utils.segmentation_utils import to_one_hot
+
+logger = get_logger(__name__)
+
+
+class AbstarctStructureLoss(_Loss, ABC):
+    """
+    Abstract computation of average structure loss between two tensors, It can support both multi-classes and binary tasks.
```
[the remark this answers is not in the document]
Comment (shanibenbaruch) [reply src/super_gradients/training/losses/structure_loss.py]: removed the word average

File: src/super_gradients/training/losses/structure_loss.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    Defined in the paper: "V-Net: Fully Convolutional Neural Networks for Volumetric Medical Image Segmentation"
```
[the remark this answers is not in the document]
Comment (shanibenbaruch) [reply src/super_gradients/training/losses/structure_loss.py]: removed in 1d42d9558a3eea9b2cc5d1aa1b9d027522e3000b

**Review by lkdci — APPROVED** [review]
LGTM

File: src/super_gradients/training/losses/structure_loss.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        if self.generalized_metric:
+            assert self.weight is None, "Cannot use structured Loss with weight classes and generalized normalization"
+            if self.eps > 1e-12:
+                logger.warning("When using GeneralizedLoss, it is recommended to use eps below 1e-12, to not affect"
+                               "small values normalized terms.")
+            if self.smooth != 0:
+                logger.warning("When using GeneralizedLoss, it is recommended to set smooth value as 0.")
+
+    @abstractmethod
+    def _calc_numerator_denominator(self, labels_one_hot, predict):
```
Comment (ofrimasad) [inline src/super_gradients/training/losses/structure_loss.py:60]: please add `-> (torch.Tensor, torch.Tensor)` to signature

**Review by ofrimasad — COMMENTED** [review]
very nice!
some minor questions and comments inline

File: src/super_gradients/training/losses/structure_loss.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+                labels_one_hot = target.unsqueeze(1)
+            else:
+                labels_one_hot = to_one_hot(target, num_classes=predict.shape[1], ignore_index=self.ignore_index)
+        else:
+            raise AssertionError(f"Mismatch of target shape: {target.size()} and prediction shape: {predict.size()},"
+                                 f" target must be [NxWxH] tensor for to_one_hot conversion"
+                                 f" or to have the same num of channels like prediction tensor")
+
+        reduce_spatial_dims = list(range(2, len(predict.shape)))
+        reduce_dims = [1] + reduce_spatial_dims if self.sum_over_batches else [0] + reduce_spatial_dims
```
[the remark this answers is not in the document]
Comment (shanibenbaruch) [reply src/super_gradients/training/losses/structure_loss.py:92]: sum_over_batches means whether to average the  *final* metric over the batch axis or classes axis. 
In this case, we want to calculate iou over batch axis if True, so we will sum the [1,2,3] channels.

File: src/super_gradients/training/losses/iou_loss.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+        super().__init__(apply_softmax=False, ignore_index=None, smooth=smooth, eps=eps, sum_over_batches=False)
+        self.apply_sigmoid = apply_sigmoid
+
+    def forward(self, predict, target):
+        if self.apply_sigmoid:
+            predict = torch.sigmoid(predict)
+        return super().forward(predict=predict, target=target)
+
+
+class GeneralizedIoULoss(IoULoss):
```
[the remark this answers is not in the document]
Comment (shanibenbaruch) [reply src/super_gradients/training/losses/iou_loss.py:67]: you'll only want to use this class for class imbalance, otherwise use regular `IoULoss`.

File: src/super_gradients/training/losses/structure_loss.py [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+    @abstractmethod
+    def _calc_numerator_denominator(self, labels_one_hot, predict):
+        """
+        All base classes must implement this function.
+        Return: 2 tensor of shape [BS, num_classes, img_width, img_height].
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def _calc_loss(self, numerator, denominator):
```
[the remark this answers is not in the document]
Comment (shanibenbaruch) [reply src/super_gradients/training/losses/structure_loss.py]: added b863160d123d7e618f80e4beab8b51472e3e64e4

File: src/super_gradients/training/losses/structure_loss.py (hunk 6 above)
Comment (shanibenbaruch) [reply src/super_gradients/training/losses/structure_loss.py]: added b863160d123d7e618f80e4beab8b51472e3e64e4

File: src/super_gradients/training/losses/structure_loss.py [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+from torch.nn.modules.loss import _Loss
+
+from super_gradients.common.abstractions.abstract_logger import get_logger
+from super_gradients.training.losses.loss_utils import apply_reduce, LossReduction
+from super_gradients.training.utils.segmentation_utils import to_one_hot
+
+logger = get_logger(__name__)
+
+
+class AbstarctStructureLoss(_Loss, ABC):
```
[the remark this answers is not in the document]
Comment (shanibenbaruch) [reply src/super_gradients/training/losses/structure_loss.py]: changed b863160d123d7e618f80e4beab8b51472e3e64e4

File: src/super_gradients/training/losses/iou_loss.py [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+        :param smooth: laplace smoothing, also known as additive smoothing. The larger smooth value is, closer the IoU
+            coefficient is to 1, which can be used as a regularization effect.
+            As mentioned in: https://github.com/pytorch/pytorch/issues/1249#issuecomment-337999895
+        :param eps: epsilon value to avoid inf.
+        """
+        super().__init__(apply_softmax=False, ignore_index=None, smooth=smooth, eps=eps, sum_over_batches=False)
+        self.apply_sigmoid = apply_sigmoid
+
+    def forward(self, predict, target):
+        if self.apply_sigmoid:
```
[the remark this answers is not in the document]
Comment (shanibenbaruch) [reply src/super_gradients/training/losses/iou_loss.py:62]: its already checked in the StructureLoss forward https://github.com/Deci-AI/super-gradients/blob/feature/ALG-277_iou_loss/src/super_gradients/training/losses/structure_loss.py#L82

```
if target.size() == predict.size():
    labels_one_hot = target
elif target.dim() == 3:       # if target tensor is in class indexes format.
    if predict.size(1) == 1 and self.ignore_index is None:    # if one class prediction task
        labels_one_hot = target.unsqueeze(1)
    else:
        labels_one_hot = to_one_hot(target, num_classes=predict.shape[1], ignore_index=self.ignore_index)
else:
    raise AssertionError(f"Mismatch of target shape: {target.size()} and prediction shape: {predict.size()},"
                         f" target must be [NxWxH] tensor for to_one_hot conversion"
                         f" or to have the same num of channels like prediction tensor")
```

File: src/super_gradients/training/losses/structure_loss.py [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+        reduce_dims = [1] + reduce_spatial_dims if self.sum_over_batches else [0] + reduce_spatial_dims
+
+        # Calculate the numerator and denominator of the chosen metric
+        numerator, denominator = self._calc_numerator_denominator(labels_one_hot, predict)
+
+        numerator = torch.sum(numerator, dim=reduce_dims)
+
+        # exclude ignore labels from denominator, false positive predicted on ignore samples are not included in
+        # total denominator.
+        if self.ignore_index is not None:
```
[the remark this answers is not in the document]
Comment (shanibenbaruch) [reply src/super_gradients/training/losses/structure_loss.py:99]: technically the nominator in both IoU and Dice is TP, which is the union between pred and one-hot target. so the pixels in the ignored index location will be zero anyway.
But for other future segmentation losses I added this commit 34251cc60debdf2e62b7b1c766401e832ceebfab. thanks!

File: src/super_gradients/training/losses/structure_loss.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (lkdci) [reply src/super_gradients/training/losses/structure_loss.py:92]: maybe a better name is `reduce_over_batches`, since it can be any reduction method sum / average / none

File: src/super_gradients/training/losses/iou_loss.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (lkdci) [reply src/super_gradients/training/losses/iou_loss.py:67]: You can always use the base `IoULoss`, this subclass is more kind of a guide of how to properly use generalized loss.
Since both nominator and denominator are normalized by `1 / sqrt(counts)` and `counts` is positive large number, i.e num of Tp pixels in the minibatch, the denominator turns into a small number. Therefore `eps` must be much smaller, and smooth is removed as suggested in the generalized class,  to prevent skewed results.

File: src/super_gradients/training/losses/structure_loss.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (shanibenbaruch) [reply src/super_gradients/training/losses/structure_loss.py]: great idea. renamed it in 90c463f9116563fde6f265e590cf42754cb8d351.

File: src/super_gradients/training/losses/iou_loss.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (ofrimasad) [reply src/super_gradients/training/losses/iou_loss.py:62]: ok

**Review by ofrimasad — APPROVED** [review]
LGTM
